### PR TITLE
Match compilation for constants

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

Add match compilation support for constants

As an aside, this reworks some name resolution stuff to add better errors

## Test plan 

A lot more tests, plus some test revisions

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/naming/translate.rs [hunk 1]
```diff
@@ -406,89 +492,99 @@ impl<'env> Context<'env> {
...
+        let function_opt = self
+            .resolve_function_opt(m, n)
+            .map(ResolvedModuleAccess::Function);
+        let constant_opt = self
+            .resolve_constant_opt(m, n)
+            .map(ResolvedModuleAccess::Constant);
+        let datatype_opt = self
+            .resolve_module_type_opt(m, n)
+            .map(ResolvedModuleAccess::Datatype);
+        match expected_access_type {
```
Comment (tzakian) [inline external-crates/move/crates/move-compiler/src/naming/translate.rs:530]: What do you think about something like this to remove some of the nesting/if-else statements?

```rust
ModuleAccess::Function => {
    function_opt.or(constant_opt).or(datatype_opt)
}
ModuleAccess::Constant => {
    constant_opt.or(datatype_opt).or(function_opt)
}
ModuleAccess::Datatype => {
    datatype_opt.or(constant_opt).or(function_opt)
}

File: external-crates/move/crates/move-compiler/src/parser/token_set.rs [hunk 2]
```diff
@@ -171,6 +171,12 @@ pub static VALUE_START_SET: Lazy<TokenSet> = Lazy::new(|| TokenSet::from(VALUE_S
 //**************************************************************************************************
 
 #[allow(dead_code)]
+impl Default for TokenSet {
```
Comment (tzakian) [inline external-crates/move/crates/move-compiler/src/parser/token_set.rs:174]: NOOOOO don't listen to clippy! This is a terrible lint rule, and IMO we should turn it off....

File: external-crates/move/crates/move-compiler/src/typing/translate.rs [hunk 3]
```diff
@@ -2255,7 +2264,31 @@ fn match_pattern_(
...
+                .insert(const_.value());
+            context.add_ability_constraint(
+                loc,
+                Some(format!(
+                    "Invalid 'copy' of value with the '{}' ability. \
+                    Literal patterns copy their values for equality checking",
+                    Ability_::Copy
+                )),
+                ty.clone(),
+                Ability_::Copy,
```
Comment (tzakian) [inline external-crates/move/crates/move-compiler/src/typing/translate.rs:2282]: Nit: formatting

**Review by tzakian — APPROVED** [review]
Overall looks good to me except for some stylistic nits.

Could you also add tests for:
* Matching on a constant and using that constant in a guard as well? E.g., a match arm like `Const if (Const == 5) => ...`
* Matching on a constant with an at-pattern, and then using that in a guard, e.g., `x @ Const if (x == 5) => ...`
* Check for order of evaluation of guards in the presence of constant matches, so something like 
```rust
let mut x = 0;
match TRUE { FALSE if ({ *x = 10; true }) => (), _ => () };
assert!(x == 0);
```

File: external-crates/move/crates/move-compiler/src/naming/translate.rs (hunk 1 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/naming/translate.rs:530]: I would have done that if I knew `or` existed -- I even looked at `std::option` looking for it

File: external-crates/move/crates/move-compiler/src/expansion/translate.rs [hunk 4]
```diff
@@ -3069,9 +3074,14 @@ fn match_pattern(context: &mut Context, sp!(loc, pat_): P::MatchPattern) -> E::M
...
                             name_value,
                         );
-                        context
-                            .env()
-                            .add_diag(diag!(Declarations::InvalidName, (name.loc, msg)));
+                        let mut diag = diag!(Declarations::InvalidName, (name.loc, msg));
+                        if is_pascal_case(&name_value) || is_upper_snake_case(&name_value) {
+                            diag.add_note(
+                                "The compiler may have failed to \
+                                          resolve this constant's name",
```
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/expansion/translate.rs:3090]: cargo fmt strikes again!

**Comment by cgswords** [comment]
Added a few more tests as requested, plus fixed everything else up.

**Pull request merged**